cardano-submit-api/src/Cardano/TxSubmit/Web.hs

```diff
@@ -135,33 +135,53 @@ txSubmitPost
   -> ByteString
   -> Handler TxId
 txSubmitPost trace metrics (AnyConsensusModeParams cModeParams) networkId (SocketPath socketPath) txBytes = handle $ do
-    InAnyCardanoEra era tx <- readByteStringTx txBytes
-    let cMode = AnyConsensusMode $ consensusModeOnly cModeParams
-    eraInMode <- hoistMaybe
-                   (TxCmdEraConsensusModeMismatch cMode (AnyCardanoEra era))
-                   (toEraInMode era $ consensusModeOnly cModeParams)
-    let txInMode = TxInMode tx eraInMode
-        localNodeConnInfo = LocalNodeConnectInfo
-                              { localConsensusModeParams = cModeParams
-                              , localNodeNetworkId = networkId
-                              , localNodeSocketPath = socketPath
-                              }
-
-    res <- liftIO $ submitTxToNodeLocal localNodeConnInfo txInMode
-    case res of
-      Net.Tx.SubmitSuccess -> do
-        liftIO $ T.putStrLn "Transaction successfully submitted."
-        return $ getTxId (getTxBody tx)
-      Net.Tx.SubmitFail reason ->
-        case reason of
-          TxValidationErrorInMode err _eraInMode -> left . TxCmdTxSubmitError . T.pack $ show err
-          TxValidationEraMismatch mismatchErr -> left $ TxCmdTxSubmitErrorEraMismatch mismatchErr
+    let localNodeConnInfo = LocalNodeConnectInfo
+          { localConsensusModeParams = cModeParams
+          , localNodeNetworkId = networkId
+          , localNodeSocketPath = socketPath
+          }
+
+    case localConsensusMode localNodeConnInfo of
+      ByronMode{} -> do
+        case deserialiseFromCBOR AsByronTx txBytes of
+          Left err -> throwError (TxCmdTxDecodeFail err)
+          Right tx -> do
+            res <- liftIO $ submitTxToNodeLocal localNodeConnInfo (TxInMode tx ByronEraInByronMode)
+            handleResult tx res
+
+      ShelleyMode{} -> do
+        case deserialiseFromCBOR AsShelleyTx txBytes of
+          Left err -> throwError (TxCmdTxDecodeFail err)
+          Right tx -> do
+            res <- liftIO $ submitTxToNodeLocal localNodeConnInfo (TxInMode tx ShelleyEraInShelleyMode)
+            handleResult tx res
+
+      CardanoMode{} -> do
+        case deserialiseFromCBOR AsByronTx txBytes of
+          Left err -> case deserialiseFromCBOR AsShelleyTx txBytes of
+            Left _ -> throwError (TxCmdTxDecodeFail err)
+            Right tx -> do
+              res <- liftIO $ submitTxToNodeLocal localNodeConnInfo (TxInMode tx ShelleyEraInCardanoMode)
+              handleResult tx res
+          Right tx -> do
+            res <- liftIO $ submitTxToNodeLocal localNodeConnInfo (TxInMode tx ByronEraInCardanoMode)
+            handleResult tx res
```

Comment:
We'll need to handle the modes and eras a bit differently, more like how it was done before here when the format was text envelope.

The patch above is both a bit verbose (since it pattern matches on the mode) but also misses out support for any eras after Shelley. That'll happen to work for Shelley txs, but cannot support new features in the Allegra or Mary eras.

I think the right thing to do here is to do a replacement for this bit:
```
InAnyCardanoEra era tx <- readByteStringTx txBytes
```
So a replacement for this would have the same type, but handle the raw CBOR format, rather than the text envelope.

With a replacement for that, the rest of `txSubmitPost` could remain as it was before I think.

So the current impl for text envelope is:
```
readByteStringInAnyCardanoEra
  :: ( HasTextEnvelope (thing ByronEra)
     , HasTextEnvelope (thing ShelleyEra)
     , HasTextEnvelope (thing AllegraEra)
     , HasTextEnvelope (thing MaryEra)
     )
  => (forall era. AsType era -> AsType (thing era))
  -> ByteString
  -> ExceptT TextEnvelopeError IO (InAnyCardanoEra thing)
readByteStringInAnyCardanoEra asThing = readByteStringTextEnvelopeAnyOf2
  [ FromSomeType (asThing AsByronEra)   (InAnyCardanoEra ByronEra)
  , FromSomeType (asThing AsShelleyEra) (InAnyCardanoEra ShelleyEra)
  , FromSomeType (asThing AsAllegraEra) (InAnyCardanoEra AllegraEra)
  , FromSomeType (asThing AsMaryEra)    (InAnyCardanoEra MaryEra)
  ]
```
We'll want to do something similar that tries decoding it as all the possible eras. Since with raw format we don't know what era it is supposed to be in (no envelope to tell us), I think the strategy ought to be just to try it in each era in turn from oldest to newest.

Reply:
Thanks!

Reply:
This has been fixed as suggested.